snowfakery/standard_plugins/Salesforce.py

```diff
@@ -70,9 +222,188 @@ def _parse_special_args(self, args):
+    dburl = f"sqlite:///{tempfile}"
+    with SqlDbOutputStream.from_url(dburl) as db:
+        ti = TableInfo("data")
+        ti.fields = {fieldname: None for fieldname in fieldnames}
+        db.create_or_validate_tables({"data": ti})
+        for row in results:
+            row_dict = {fieldname: result for fieldname, result in zip(fieldnames, row)}
+            db.write_row("data", row_dict)
+        db.flush()
+        db.close()
+    return tempdir, dburl
+
+
+class SOQLDataset(SalesforceConnectionMixin, DatasetPluginBase):
+    def __init__(self, *args, **kwargs):
+        self.dataset_impl = SOQLDatasetImpl(self)
+        super().__init__(*args, **kwargs)
+
+
+class SalesforceQuery(SalesforceConnectionMixin, SnowfakeryPlugin):
+    class Functions:
+        @property
+        def _sf_connection(self):
+            return self.context.plugin.sf_connection
+
+        def random_record(self, *args, fields="Id", where=None, **kwargs):
+            """Query a random record."""
+            context_vars = self.context.context_vars()
+            context_vars.setdefault("count_query_cache", {})
+
+            # "from" has to be handled separately because its a Python keyword
+            query_from = self._parse_from_from_args(args, kwargs)
+
+            if not query_from:
+                raise ValueError("Need to specify a table as 'from' argument")
+
+            # TODO: Test WHERE
+            where_clause = f" WHERE {where}" if where else ""
+            count_query = f"SELECT count() FROM {query_from}{where_clause}"
+            count_result = self._sf_connection.query(count_query)
+
+            count = count_result["totalSize"]
+            mx = min(count, MAX_SALESFORCE_OFFSET)
+            context_vars["count_query_cache"][count_query] = mx
+            rand_offset = randrange(0, mx)
+            query = f"SELECT {fields} FROM {query_from}{where_clause} LIMIT 1 OFFSET {rand_offset}"
+            # todo: use CompositeParallelSalesforce to cache 200 at a time
+            return self._sf_connection.query_single_record(query)
+
+        def find_record(self, *args, fields="Id", where=None, **kwargs):
+            """Find a particular record"""
+            query_from = self._parse_from_from_args(args, kwargs)
+            where_clause = f" WHERE {where}" if where else ""
+            query = f"SELECT {fields} FROM {query_from}{where_clause} LIMIT 1"
+            return self._sf_connection.query_single_record(query)
+
+        def _parse_from_from_args(self, args, kwargs):
+            if kwargs:
+                query_from = kwargs.pop("from", None)
+
+                if kwargs:
+                    raise ValueError(f"Unknown arguments: {tuple(kwargs.keys())}")
+            elif args:
+                if len(args) != 1 or not isinstance(args[0], str):
+                    raise ValueError(f"Only one string argument allowed, not: {args}")
+                query_from = args[0]
+
+            return query_from
```

Comment:
This module only has ~63% test coverage. (At least on coveralls...is that because of not running the integration tests in CI yet?)